Transform pixel track fitter, filter, and cleaner to ED or ES products

makortel opened the pull request:
This PR contains intermediate developments before migrating `PixelTrackProducer` to the new seeding framework of #16635. The point is to make all other components than the hit ntuplet generation to support `fillDescriptions()`. This is achieved by making fitters and filters EDProducts (because at least one of the algorithms depend on event data) and making cleaners ESProducts (because none of the cleaners need event data).

I have tried to migrate all uses in CMSSW, but I can't exclude of missing something.

The next step is to migrate `PixelTrackProducer` to the new seeding framework (#16635), and fully migrate HLT (with `customizeHLTforCMSSW.py`).

Tested in CMSSW_9_0_X_2016-11-27-2300 (rebased on top of CMSSW_9_0_X_2016-12-04-2300), no changes expected.

@rovere @VinInn @felicepantaleo @mtosi @JanFSchulte

makortel committed:
Simplify PixelTrackFilter interface

No need to have three operator()'s, as the hits are always available
and the implementations needing TrackerTopology can just fetch it by
themselves.

makortel committed:
Rename PixelTrackFilter -> PixelTrackFilterBase

I intend to use PixelTrackFilter as the name of an EDProduct type soon.

makortel committed:
Migrate PixelTrackFilters from helper plugins to EDProducts

makortel committed:
PixelFitter -> PixelFitterBase

makortel committed:
Migrate PixelFitters from helper plugins to EDProducts

These need to be EDProducts instead of ESProducts because
KFBasedPixelFitter needs event data.

makortel committed:
Migrate PixelTrackCleaners to ESProducts

These can be ESProducts as none of the cleaners need event data.

makortel committed:
Remove last bits of use of PixelTrackCleaner from TSGFromL1Muon

makortel committed:
Deliver TrackerTopology via ESProducers to TrackCleaner

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for CMSSW_9_0_X.

It involves the following packages:

Calibration/IsolatedParticles
DQM/BeamMonitor
DQM/Integration
FastSimulation/HighLevelTrigger
FastSimulation/Tracking
HLTrigger/Configuration
RecoHI/HiJetAlgos
RecoHI/HiTracking
RecoMuon/L3MuonProducer
RecoMuon/TrackerSeedGenerator
RecoPixelVertexing/Configuration
RecoPixelVertexing/PixelLowPtUtilities
RecoPixelVertexing/PixelTrackFitting
RecoTracker/Record


@perrotta, @cmsbuild, @civanch, @lveldere, @silviodonato, @cvuosalo, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @Martin-Grunewald, @franzoni, @ghellwig, @cerminar, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @yslai, @felicepantaleo, @abbiendi, @Martin-Grunewald, @threus, @geoff-smith, @battibass, @jhgoh, @HuguesBrun, @trocino, @yetkinyilmaz, @GiacomoSguazzoni, @rovere, @VinInn, @bellan, @tocheng, @jalimena, @mschrode, @dgulhan, @echapon, @batinkov, @mandrenguyen, @jazzitup, @calderona, @yenjie, @kurtejung, @gpetruc, @matt-komm, @rociovilar, @bachtis this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 7b054b636944b8e04558e3db5cfef5f063dc2d54

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16652/summary.html

I found follow errors while testing this PR

**Failed tests:  AddOn**

* **AddOn**:

I found errors in the following addon tests:

cmsRun /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-11-28-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Tue Nov 29 16:18:12 2016-date Tue Nov 29 16:17:10 2016 s - exit: 23040
cmsDriver.py RelVal -s HLT:HIon,RAW2DIGI,L1Reco,RECO --data --scenario=HeavyIons -n 10 --conditions auto:run2_data_HIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016,Run2_HI --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_HIon_DATA.root --fileout file:RelVal_Raw_HIon_DATA_HLT_RECO.root : FAILED - time: date Tue Nov 29 16:18:12 2016-date Tue Nov 29 16:17:10 2016 s - exit: 256
cmsRun /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-11-28-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Tue Nov 29 16:23:07 2016-date Tue Nov 29 16:17:31 2016 s - exit: 23040
cmsDriver.py RelVal -s HLT:HIon,RAW2DIGI,L1Reco,RECO --mc --scenario=HeavyIons -n 10 --conditions auto:run2_mc_HIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016,Run2_HI --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_HIon_MC.root --fileout file:RelVal_Raw_HIon_MC_HLT_RECO.root : FAILED - time: date Tue Nov 29 16:23:07 2016-date Tue Nov 29 16:17:31 2016 s - exit: 256

cmsbuild commented:
Comparison job queued.

makortel committed:
Fix typo

makortel commented:
@cmsbuild, please test

(trivial typo; but are there instructions somewhere how I could run the "AddOn" tests myself?)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #16792 was updated. @perrotta, @cmsbuild, @civanch, @lveldere, @silviodonato, @cvuosalo, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @Martin-Grunewald, @franzoni, @ghellwig, @cerminar, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

Martin-Grunewald commented:
```addOnTests.py```

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16652/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

cmsbuild commented:
+1
Tested at: 6bd9a5c23fd7e207a098e4a4b3a78dddf2d68a85
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16655/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16655/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

makortel commented:
@Martin-Grunewald Thanks, very simple.

@smuzaffar Would it make sense to give the recipe to re-run the AddOn tests in the cms-bot message in case the tests failed (or in the test result page)?

smuzaffar commented:
@makortel , I will update cms-bot to add the recipe to run addon test.

makortel commented:
@smuzaffar Thanks!

Martin-Grunewald commented on HLTrigger/Configuration/python/customizeHLTforCMSSW.py:
```diff
@@ -90,6 +168,7 @@ def customizeHLTforCMSSW(process, menuType="GRun"):
         process = customiseFor15440(process)
         process = customiseFor15499(process)
         process = customiseFor16569(process)
+        process = customiseFor16792(process)
```
Yes, please.

makortel committed:
Move customiseFor16792 call under 'if cmssw >= 9_0'

makortel committed:
Remove useFilterWithES parameter in customizeHLTforCMSSW

cmsbuild commented:
Pull request #16792 was updated. @perrotta, @cmsbuild, @civanch, @lveldere, @silviodonato, @cvuosalo, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @Martin-Grunewald, @franzoni, @ghellwig, @cerminar, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

makortel commented:
@cmsbuild, please test

Moved the HLT customize function call under `if cmsswVersion >= "CMSSW_9_0":`, and added customize code to remove `useFilterWithES` parameter from `PixelTrackProducer` (it became obsolete; removing it now means that I don't have to deal with it in the "next PR" where I'll add `fillDescriptions()` to `PixelTrackProducer`).

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: c2b82d31e90fc5736b422df4e5b1baaca37579da
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16686/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16686/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

makortel commented:
@cmsbuild, please test

Rebased on top of CMSSW_9_0_X_2016-12-04-2300.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #16792 was updated. @perrotta, @cmsbuild, @civanch, @lveldere, @silviodonato, @cvuosalo, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @Martin-Grunewald, @franzoni, @ghellwig, @cerminar, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 1cf669ccb0c3831e98f3f9d15dcc1cada808abfb
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16782/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16782/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

makortel commented:
@slava77 Regarding your question if this PR would conflict with #16885 (it was that, right?), the answer is **no**. They modify different files.

slava77 commented:
On 12/8/16 4:34 AM, Matti Kortelainen wrote:
> @slava77 <https://github.com/slava77> Regarding your question if this PR
> would conflict with #16885 <https://github.com/cms-sw/cmssw/pull/16885>
> (it was that, right?), the answer is *no*. They modify different files.
> 

Thank you for the confirmation.


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16792#issuecomment-265729869>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcblPgUzgIDsar7Y4xO2fsMFz0UomUks5rF_k-gaJpZM4K_AR1>.
>

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/16792 1cf669c
- mostly technical changes moving filter/fitter/cleaner tools internal to pixel track producer  modules  to ES or ED products
- jenkins tests pass and comparisons with baseline show no relevant differences
    - some HLT HeavyFlavor Upsilon plots (e.g. /HLT_Mu7p5_Track3p5_Upsilon_v) have changed in a way that suggests some random issues, like the same histogram written to by different modules

makortel commented:
> some HLT HeavyFlavor Upsilon plots (e.g. /HLT_Mu7p5_Track3p5_Upsilon_v) have changed in a way that suggests some random issues, like the same histogram written to by different modules

Do we have a list of "known" DQM histograms with reproducibility issues? If not could we make one (and maybe get the issues fixed at some point)? If @slava77 is referring to e.g. 25202.0, I've just wasted some hours and hairs with this histogram when developing the next step after this PR.

slava77 commented:
We regularly ignore the following pattern

/HLT_IsoTkMu\|/HLT_OldIsoTkMu

The Upsilon plot hasn't come up before,
however, given that the title of the histogram changed, I noted that it's something random (diff is done based on the histogram name)

<img width="967" alt="wf25202_upsilon_hlt" src="https://cloud.githubusercontent.com/assets/4676718/21026981/c4f8fbc6-bd43-11e6-91e8-2355e2ec2b74.png">

makortel commented:
Actually this might be related to this PR. `hltPixelTracksL3MuonFilter` is a module added in `customizeHLTforCMSSW` of type `PixelTrackFilterByKinematicsProducer`, whereas `hltMu7p5Track2UpsilonTrackMassFiltered` is of type `HLTMuonTrackMassFilter`. I wonder how this change enters up in a DQM histogram though. Ok, taking a quick look on the code, this logic
https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc#L156
looks like a plausible cause for that.

Good ideas how to move on?  A quick&dirty solution would be to avoid postfix "Filter" here
https://github.com/cms-sw/cmssw/pull/16792/files#diff-5b29e53ee0c1d396908444c2711d39f6R114
but e.g. "Filtr" (or "filter" with lowecase "f"), and leave the validation logic to be enhanced by the authors (assuming HLT prefers to possibly use the "Filter" as part of the pixel tracking filters in the module labels). Or leave it as it is? (I'll anyway run a test overnight changing the postfix to something else, to confirm the diagnosis if nothing else)

(personally I feel checking only the module labels and not taking into account the C++ types in the HLT path "parsing" is fragile...)

slava77 commented:
On 12/8/16 1:26 PM, Matti Kortelainen wrote:
> Actually this might be related to this PR. |hltPixelTracksL3MuonFilter|
> is a module added in |customizeHLTforCMSSW| of type
> |PixelTrackFilterByKinematicsProducer|, whereas
> |hltMu7p5Track2UpsilonTrackMassFiltered| is of type
> |HLTMuonTrackMassFilter|. I wonder how this change enters up in a DQM
> histogram though. Ok, taking a quick look on the code, this logic
> https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc#L156
> looks like a plausible cause for that.

I removed my +1


> 
> Good ideas how to move on? A quick&dirty solution would be to avoid
> postfix "Filter" here

I think it's better to add a check of the module type in the
HeavyFlavorValidation

		--slava


> https://github.com/cms-sw/cmssw/pull/16792/files#diff-5b29e53ee0c1d396908444c2711d39f6R114
> but e.g. "Filtr" (or "filter" with lowecase "f"), and leave the
> validation logic to be enhanced by the authors (assuming HLT prefers to
> possibly use the "Filter" as part of the pixel tracking filters in the
> module labels). Or leave it as it is? (I'll anyway run a test overnight
> changing the postfix to something else, to confirm the diagnosis if
> nothing else)
> 
> (personally I feel checking only the module labels and not taking into
> account the C++ types in the HLT path "parsing" is fragile...)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16792#issuecomment-265860235>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcboVPb8wptdEVRbjdf5qetF_aEOqqks5rGHX6gaJpZM4K_AR1>.
>

cmsbuild commented:
Pull request has been put on hold by  @slava77
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

makortel commented:
> >  Good ideas how to move on? A quick&dirty solution would be to avoid
> >  postfix "Filter" here
> 
> I think it's better to add a check of the module type in the
> HeavyFlavorValidation

I have no idea what other type of modules the `HeavyFlavorValidation` would like to monitor than `HLTMuonTrackMassFilter`, so without more information I don't know how exactly to fix.

makortel commented:
Or then I make it to ignore all the added `PixelTrackFIlter` producers.

slava77 commented:
On 12/8/16 2:13 PM, Matti Kortelainen wrote:
>         Good ideas how to move on? A quick&dirty solution would be to avoid
>         postfix "Filter" here
> 
>     I think it's better to add a check of the module type in the
>     HeavyFlavorValidation
> 
> I have no idea what other type of modules the |HeavyFlavorValidation|
> would like to monitor than |HLTMuonTrackMassFilter|, so without more
> information I don't know how exactly to fix.

"EDFilter"  == moduleEDMType(module)

this is used in a bunch of other DQM modules.



> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/16792#issuecomment-265871604>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbsCbOBq4Wm3bfmjYljB-180BarkUks5rGIELgaJpZM4K_AR1>.
>

makortel committed:
Require filters of interest in HeavyFlavorValidation to have EDM type of EDFilter

cmsbuild commented:
Pull request #16792 was updated. @perrotta, @cmsbuild, @civanch, @lveldere, @silviodonato, @cvuosalo, @arunhep, @ssekmen, @mdhildreth, @dmitrijus, @Martin-Grunewald, @franzoni, @ghellwig, @cerminar, @slava77, @fwyzard, @mmusich, @vanbesien, @davidlange6 can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented:
@slava77 
> "EDFilter"  == moduleEDMType(module)

That seemed to do the trick (and to be generic-enough), thanks.

cmsbuild commented:
+1
Tested at: 5e6513c3c8beef6c38e792ae644dfac51055e34f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16889/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16889/summary.html

slava77 commented:
@makortel @boudoul @veszpv 
any idea why data (1000.0 and 136.731) MechanicalView plots go in and out at least in recent IBs?

This PR baseline vs PR bring the plots in:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_0_X_2016-12-08-2300+16792/17469/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/AlCaReco_SiStrip.html

Unrelated to tracking PR #16909, same as this PR:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_0_X_2016-12-08-2300+16909/17480/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/AlCaReco_SiStrip.html

Some other unrelated PR (tested in the same IB) shows no diffs
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_0_X_2016-12-08-2300+16943/17473/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/AlCaReco_SiStrip.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/16792 5e6513c
- the issue found in HLT validation after the last signoff  https://github.com/cms-sw/cmssw/pull/16792#issuecomment-265785056  was addressed
- jenkins tests pass and comparisons with baseline show no relevant differences (the diff in Upsilon plots in HLT is gone)

slava77 commented:
unhold

slava77 commented:
ping
we need DQM, HLT, AlCa and FastSim signatures here

Martin-Grunewald commented:
please test
(HLT menu has changed since in 90X)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5e6513c3c8beef6c38e792ae644dfac51055e34f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16990/summary.html

cmsbuild commented:
Comparison job queued.

Martin-Grunewald commented:
+1

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16792/16990/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * 20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D7_GenSimHLBeamSpotFull14+DigiFull_2023D7+RecoFullGlobal_2023D7+HARVESTFullGlobal_2023D7
   * 23234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D8_GenSimHLBeamSpotFull14+DigiFull_2023D8+RecoFullGlobal_2023D8+HARVESTFullGlobal_2023D8

mmusich commented:
+1

makortel commented:
For the follow-up PR I'll wait until either #17041 or #17042 gets merged (because I have to migrate the customize function, and doing anything there now would just create conflicts).

fwyzard commented:
By the way @makortel , which approach do you prefer,  #17041 or #17042 ?
I'm trying to figure out with @Martin-Grunewald and @silviodonato which one to choose, feel free to chime in.

makortel commented:
@fwyzard I can certainly share my experience from offline tracking, where are you having the discussion?(this PR is likely not the right forum)
